### ISSUE TITLE
PROXY_TO_PTHREAD support for Embind

### ISSUE THREAD

**Issue by kleisauke** [emscripten-core/emscripten#9847]
I'm creating JavaScript bindings (using Embind) for a C/C++ library that uses a lot of threading features. The bindings are compiled with these Emscripten compiler flags:
```
-s USE_PTHREADS=1 -s PROXY_TO_PTHREAD=1 -s ALLOW_BLOCKING_ON_MAIN_THREAD=0
```

Which moves the application’s `main()`  to a pthread and disallows to block the main browser thread.

I noticed that when an Embind wrapped function (that will eventually use `pthread_join` or `pthread_cond_wait`) is invoked on the main browser thread, an error is immediately thrown. This indicates that the wrapped function is executed on the calling thread, and _not_ on the pthread running the original `main()`.

Is it possible to execute these functions on the pthread running the original `main()`?

<details><summary>Minimal reproduction</summary>

```cpp
// answer.cpp

#include <emscripten.h>
#include <emscripten/bind.h>
#include <emscripten/val.h>

#include <future>
#include <thread>

using namespace emscripten;

int main(int argc, char *argv[]) {
    // Ensure we don't exit the runtime, see:
    // https://github.com/emscripten-core/emscripten/issues/8202
    emscripten_exit_with_live_runtime();

    return 0;
}

int answer() {
    // An arbitrary but carefully chosen number
    return 42;
}

int answerAsync() {
    std::future<int> future = std::async(answer);
    return future.get();
}

EMSCRIPTEN_BINDINGS(my_module) {
    function("answer", &answerAsync);
}
```

Compile with:
```bash
# Note: we set PTHREAD_POOL_SIZE to 2, otherwise the program hangs if the number of
# active threads exceed PTHREAD_POOL_SIZE. See:
# https://github.com/emscripten-core/emscripten/issues/8988
emcc answer.cpp \
  -std=c++11 \
  --bind \
  -s USE_PTHREADS=1 \
  -s PROXY_TO_PTHREAD=1 \
  -s ALLOW_BLOCKING_ON_MAIN_THREAD=0 \
  -s PTHREAD_POOL_SIZE=2 \
  -o test.html
```

Run `Module.answer()` in the developer tools console and notice that the `Blocking on the main thread is not allowed by default` error is thrown.
</details>

**Comment by kripken** [comment]
This is surprising - I don't think we proxy embind calls to the main thread. Perhaps the full call stack leading to that error shows some proxying code, that could explain things? (Or maybe I've forgotten something here...)

cc @jgravelle-google

**Comment by kleisauke** [comment]
Here's the stacktrace:
```
test.js:9206 Uncaught RuntimeError: abort(Blocking on the main thread is not allowed by default. See https://emscripten.org/docs/porting/pthreads.html#blocking-on-the-main-browser-thread) at Error
    at jsStackTrace (http://localhost:5000/test.js:1932:17)
    at stackTrace (http://localhost:5000/test.js:1949:16)
    at abort (http://localhost:5000/test.js:1700:44)
    at _emscripten_check_blocking_allowed (http://localhost:5000/test.js:5926:7)
    at __pthread_cond_timedwait (wasm-function[1261]:0x3af2f)
    at std::__2::__libcpp_condvar_wait(pthread_cond_t*, pthread_mutex_t*) (wasm-function[112]:0x2c6b)
    at dynCall_iii (wasm-function[1307]:0x3d430)
    at Module.dynCall_iii (http://localhost:5000/test.js:8988:39)
    at invoke_iii (http://localhost:5000/test.js:9181:12)
    at std::__2::condition_variable::wait(std::__2::unique_lock<std::__2::mutex>&) (wasm-function[110]:0x2bfa)
    at abort (http://localhost:5000/test.js:1706:9)
    at _emscripten_check_blocking_allowed (http://localhost:5000/test.js:5926:7)
    at __pthread_cond_timedwait (wasm-function[1261]:0x3af2f)
    at std::__2::__libcpp_condvar_wait(pthread_cond_t*, pthread_mutex_t*) (wasm-function[112]:0x2c6b)
    at dynCall_iii (wasm-function[1307]:0x3d430)
    at Module.dynCall_iii (http://localhost:5000/test.js:8988:39)
    at invoke_iii (http://localhost:5000/test.js:9181:12)
    at std::__2::condition_variable::wait(std::__2::unique_lock<std::__2::mutex>&) (wasm-function[110]:0x2bfa)
    at std::__2::__assoc_sub_state::__sub_wait(std::__2::unique_lock<std::__2::mutex>&) (wasm-function[130]:0x2fb7)
    at dynCall_vii (wasm-function[1318]:0x3d502)
```

Notice that it will call the C function pointer from JS on the main browser thread. I had a quick go with patching the [`makeDynCaller` function in `$embind__requireFunction`](https://github.com/emscripten-core/emscripten/blob/1abc0c76764e33e1f14fe7a8e5f44eedec8ddb41/src/embind/embind.js#L1051-L1074), see:

<details><summary>concept9847.js</summary>

```js
// concept9847.js
// build with: --js-library concept9847.js

// --js-library must go after --bind
if (!LibraryManager.library['$embind__requireFunction']) {
    throw Error("embind.js didn't run yet");
}

mergeInto(LibraryManager.library, {
    $embind__requireFunction: function (signature, rawFunction) {
        signature = readLatin1String(signature);

        function makeDynCallerProxy() {
            var args = [];
            for (var i = 1; i < signature.length; ++i) {
                args.push('a' + i);
            }

            // TODO: Make it work for other signatures as well. Perhaps something like this?:
            /*var sig;
            switch (signature) {
                case 'v':
                    sig = {{{ cDefine('EM_FUNC_SIG_V') }}};
                    break;
                case 'vi':
                    sig = {{{ cDefine('EM_FUNC_SIG_VI') }}};
                    break;
                case 'vii':
                    sig = {{{ cDefine('EM_FUNC_SIG_VII') }}};
                    break;
                case 'viii':
                    sig = {{{ cDefine('EM_FUNC_SIG_VIII') }}};
                    break;
                case 'i':
                    sig = {{{ cDefine('EM_FUNC_SIG_I') }}};
                    break;
                case 'ii':
                    sig = {{{ cDefine('EM_FUNC_SIG_II') }}};
                    break;
                case 'iii':
                    sig = {{{ cDefine('EM_FUNC_SIG_III') }}};
                    break;
                case 'iiii':
                    sig = {{{ cDefine('EM_FUNC_SIG_IIII') }}};
                    break;
                default:
                    throwBindingError("No dynCall invoker for signature: " + signature);
            }*/

            var name = 'dynCall_' + signature + '_' + rawFunction;
            var body = 'return function ' + name + '(' + args.join(', ') + ') {\n';
            // body += '    var sig = '+ sig + ';\n';
            body += '    var stackTop = stackSave();\n';
            // TODO: These parameters are hardcoded, obviously we want something smarter here.
            body += '    var varargs = stackAlloc(4);\n';
            body += '    {{{ makeSetValue('varargs', 0, 'a1', 'i32')}}};\n';
            // TODO: 5414848 = is the result of pthread_self() in the proxied main (`__emscripten_thread_main`)
            body += '     _emscripten_async_queue_on_thread_(5414848, {{{ cDefine('EM_FUNC_SIG_II') }}}, rawFunction, 0, varargs);\n';
            body += '    stackRestore(stackTop);\n';
            // TODO: Wait with `emscripten_wait_for_call_v` and return result.
            body    += '};\n';

            return (new Function('rawFunction', body))(rawFunction);
        }

        var fp;
        if (Module['FUNCTION_TABLE_' + signature] !== undefined) {
            fp = Module['FUNCTION_TABLE_' + signature][rawFunction];
        } else if (typeof FUNCTION_TABLE !== "undefined") {
            fp = FUNCTION_TABLE[rawFunction];
        } else {
            fp = makeDynCallerProxy();
        }

        if (typeof fp !== "function") {
            throwBindingError("unknown function pointer with signature " + signature + ": " + rawFunction);
        }
        return fp;
    },
});
```
</details>

With this patch, `answerAsync()` is called on the [proxied main (`__emscripten_thread_main`)](https://github.com/emscripten-core/emscripten/blob/3bb812356bceb3fee85eef2fe45675fff77d403e/system/lib/pthread/library_pthread.c#L927-L932) and therefore won't block the main browser thread. Obviously it will not return any result (since I didn't wait with [`emscripten_wait_for_call_v`](https://github.com/emscripten-core/emscripten/blob/3bb812356bceb3fee85eef2fe45675fff77d403e/system/lib/pthread/library_pthread.c#L414)) and it only works with this function signature (I've hardcoded some things).

Unfortunately, I couldn't use the handy [`emscripten_sync_run_in_main_thread_*` methods](https://github.com/emscripten-core/emscripten/blob/1abc0c76764e33e1f14fe7a8e5f44eedec8ddb41/system/lib/pthread/library_pthread.c#L564-L675) here, since these functions will run on [`emscripten_main_browser_thread_id`](https://github.com/emscripten-core/emscripten/blob/1abc0c76764e33e1f14fe7a8e5f44eedec8ddb41/system/lib/pthread/library_pthread.c#L553-L555) and not on the proxied main.

This could be a bit of a niche issue. Hopefully I've clarified a couple of things.

**Comment by kleisauke** [comment]
Here is an improved workaround, which seems to work with this particular example:

<details><summary>Workaround</summary>

```cpp
// answer.cpp

#include <emscripten.h>
#include <emscripten/bind.h>
#include <emscripten/threading.h>
#include <emscripten/val.h>

#include <cmath>
#include <future>
#include <thread>

using namespace emscripten;

static pthread_t main_application_thread_id_ = 0;

int main(int argc, char *argv[]) {
    main_application_thread_id_ = pthread_self();

    // Ensure we don't exit the runtime, see:
    // https://github.com/emscripten-core/emscripten/issues/8202
    EM_ASM(noExitRuntime = true);

    return 0;
}

extern "C" int EMSCRIPTEN_KEEPALIVE run_in_main_application_thread(
    EM_FUNC_SIGNATURE sig, void *func_ptr, int *buffer) {
    int numArguments = EM_FUNC_SIG_NUM_FUNC_ARGUMENTS(sig);
    em_queued_call q = {static_cast<int>(sig), func_ptr};

    for (int i = 0; i < numArguments; i++) {
        q.args[i].i = buffer[i];
    }

    emscripten_async_queue_call_on_thread(main_application_thread_id_, &q);

    emscripten_wait_for_call_v(&q, INFINITY);

    return q.returnValue.i;
}

int answer() {
    return 42;
}

int answerAsync() {
    std::future<int> future = std::async(answer);
    return future.get();
}

EMSCRIPTEN_BINDINGS(my_module) {
    function("answer", &answerAsync);
}
```

```js
// workaround9847.js

// --js-library must go after --bind
if (!LibraryManager.library['$embind__requireFunction']) {
    throw Error("embind.js didn't run yet");
}

mergeInto(LibraryManager.library, {
    $embind__requireFunction: function (signature, rawFunction) {
        signature = readLatin1String(signature);

        function makeDynCallerProxy() {
            var returnValue = {
                'v': 0,
                'i': 1 << 29,
                'f': 3 << 29,
                'd': 4 << 29,
            };
            var param = {
                'i': 0,
                'f': 2 << 29,
                'd': 3 << 29,
            };

            var sig = returnValue[signature[0]] | ((signature.length - 1) << 25);
            for (var i = 1; i < signature.length; ++i) {
                sig |= param[signature[i]] << (2 * (i - 1));
            }

            return function() {
                var stack = stackSave();
                var args = stackAlloc(arguments.length * 4);
                var b = args >> 2;
                for (var i = 0; i < arguments.length; i++) {
                    HEAP32[b + i] = arguments[i];
                }
                var ret = _run_in_main_application_thread(sig, rawFunction, args);
                stackRestore(stack);
                return ret;
            };
        }

        var fp;
        if (Module['FUNCTION_TABLE_' + signature] !== undefined) {
            fp = Module['FUNCTION_TABLE_' + signature][rawFunction];
        } else if (typeof FUNCTION_TABLE !== "undefined") {
            fp = FUNCTION_TABLE[rawFunction];
        } else {
            fp = makeDynCallerProxy();
        }

        if (typeof fp !== "function") {
            throwBindingError("unknown function pointer with signature " + signature + ": " + rawFunction);
        }
        return fp;
    },
});
```

Compile with:
```bash
emcc answer.cpp \
  -std=c++11 \
  --bind \
  --js-library workaround9847.js \
  -s USE_PTHREADS=1 \
  -s PROXY_TO_PTHREAD=1 \
  -s ALLOW_BLOCKING_ON_MAIN_THREAD=0 \
  -s PTHREAD_POOL_SIZE=2 \
  -s EXPORTED_FUNCTIONS='["_main", "_run_in_main_application_thread"]' \
  -o test.html
```

*This requires that the [`emscripten_async_queue_call_on_thread`](https://github.com/emscripten-core/emscripten/blob/3bb812356bceb3fee85eef2fe45675fff77d403e/system/lib/pthread/library_pthread.c#L454) function is part of the public API.
</details>

Unfortunately this workaround doesn't work with the bindings that I'm working on, it seems to fail on [this assert](
https://github.com/emscripten-core/emscripten/blob/3bb812356bceb3fee85eef2fe45675fff77d403e/system/lib/pthread/library_pthread.c#L383) when it's calling [`emscripten_current_thread_process_queued_calls`](https://github.com/emscripten-core/emscripten/blob/3bb812356bceb3fee85eef2fe45675fff77d403e/system/lib/pthread/library_pthread.c#L677) in [a worker](https://github.com/emscripten-core/emscripten/blob/b0eaaa5f1aee2e68fea55fd84c87d7db59646371/src/worker.js#L255). I'll look into the cause of this and try to make a minimal reproduction.

**Comment by kripken** [comment]
Interesting. I'm not sure what's going on here, as I see no proxying in that stack trace, weird...

So this may be an embind bug, but I don't see any proxying in the embind code.

Can you perhaps add some debug logging in the various functions, to see where it proxies from one thread to the other?

**Comment by kleisauke** [comment]
I updated a few comments and the issue title, since I mixed up the `--proxy-to-worker` and the `-s PROXY_TO_PTHREAD=1` linker flags.  Also the workaround is getting a bit too big, so I moved it to a separate gist:
https://gist.github.com/kleisauke/bb1fdcf0e282234376532de3b423a931

> Can you perhaps add some debug logging in the various functions, to see where it proxies from one thread to the other?

Sure, which functions should I debug specifically? If you want, I can also provide the output of `-s LIBRARY_DEBUG=1`.

**Comment by kripken** [comment]
The goal would be to get a stack trace showing some embind code that calls proxying code, or at least the proxying code itself, with maybe something else calling it.

Alternatively, if you can reduce this to a simple testcase with instructions, that would help too - I find it hard to follow the "workaround" sample you provide. However, perhaps @jgravelle-google who knows more about embind can find time to look at this, as the answer may be something trivial I am missing.

**Comment by kleisauke** [comment]
I've created a minimal reproducible testcase. See commit https://github.com/kleisauke/emscripten/commit/240b43bb70697406df4b0100a0c4adc56364fe9d.

To reproduce:
```bash
git clone --single-branch --branch testcase-9847 https://github.com/kleisauke/emscripten.git testcase-9847
cd testcase-9847
python tests/runner.py browser.test_pthread_main_thread_blocking_embind
```

Hope this helps.

**Comment by kleisauke** [comment]
I just came across #9910, which seems to be a much more convenient solution than moving all embind `dynCall`s to a separate pthread.

**Comment by kleisauke** [comment]
`PROXY_TO_PTHREAD` support for Embind wrapped functions is odd, so I'll close this issue for now.

The issue I had is that the error produced by the `-s ALLOW_BLOCKING_ON_MAIN_THREAD=0` compiler flag always occurs, even when compiling with `-s PTHREAD_POOL_SIZE=N` where `N` > 0. See for example this test case: https://github.com/kleisauke/emscripten/commit/6fd7e6b346c45666baa9042ace56a337ebfb11fa.

Ideally, the error in this test case should only occur if the `PTHREAD_POOL_SIZE` compiler flag is omitted because it works properly (i.e. no deadlocks) when compiling with a predefined pool (#8988 might be relevant here).
